pkg/pgmodel/sql_ingest.go

```diff
@@ -719,44 +768,107 @@ func (h *insertHandler) setSeriesIds(sampleInfos []samplesInfo) (string, error)
 		lastSeenLabel = curr.labels
 	}
 
+	if numSQLFunctionCalls != len(batchSeries) {
+		return "", epoch, fmt.Errorf("unexpected difference in numQueries and batchSeries")
+	}
+
 	br, err := h.conn.SendBatch(context.Background(), batch)
 	if err != nil {
-		return "", err
+		return "", epoch, err
 	}
 	defer br.Close()
 
-	if numSQLFunctionCalls != len(batchSeries) {
-		return "", fmt.Errorf("unexpected difference in numQueries and batchSeries")
+	// BEGIN;
+	_, err = br.Exec()
+	if err != nil {
+		return "", epoch, err
+	}
+
+	var newEpoch int64
+	row := br.QueryRow()
+	err = row.Scan(&newEpoch)
+	if err != nil {
+		return "", epoch, err
+	}
+	if epoch == -1 || newEpoch < epoch {
+		epoch = newEpoch
+	}
+
+	// COMMIT;
+	_, err = br.Exec()
+	if err != nil {
+		return "", epoch, err
 	}
 
 	var tableName string
 	for i := 0; i < numSQLFunctionCalls; i++ {
+		// BEGIN;
 		_, err = br.Exec()
 		if err != nil {
-			return "", err
+			return "", epoch, err
 		}
-		row := br.QueryRow()
 
 		var id SeriesID
+		row = br.QueryRow()
 		err = row.Scan(&tableName, &id)
 		if err != nil {
-			return "", err
+			return "", epoch, err
 		}
-		h.seriesCache[batchSeries[i][0].labels.String()] = id
+		seriesString := batchSeries[i][0].labels.String()
+		h.seriesCache[seriesString] = id
 		for _, lsi := range batchSeries[i] {
 			lsi.seriesID = id
 		}
+
+		// COMMIT;
 		_, err = br.Exec()
 		if err != nil {
-			return "", err
+			return "", epoch, err
 		}
 	}
 
-	return tableName, nil
+	return tableName, epoch, nil
+}
+
+func (h *insertHandler) refreshSeriesCache() {
+	newEpoch, err := h.getServerEpoch()
+	if err != nil {
+		// we don't have any great place to report this error, and if the
+		// connection recovers we can still make progress, so we'll just log it
+		// and continue execution
+		msg := fmt.Sprintf("error refreshing the series cache for %s", h.metricTableName)
+		log.Error("msg", msg, "metric_table", h.metricTableName, "err", err)
+		// Trash the cache just in case an epoch change occured, seems safer
+		h.seriesCache = map[string]SeriesID{}
```

Comment:
should we also add h.seriesCacheEpoch = -1 here?

Reply:
ehh, why not